### ISSUE TITLE
 [PhpUnitBridge] Don't show deprecation warnings when getting private services from test client container

### ISSUE THREAD

**Issue by arderyp** [symfony/symfony#27037]
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| BC Break report? | no
| RFC?             | ?
| Symfony version  | 4.1.x

## Related
- https://github.com/symfony/symfony/pull/26499
- https://github.com/symfony/symfony/pull/24418
- https://github.com/symfony/symfony/issues/23311

The issues above demonstrate the newly committed functionality whereby we can freely access private services from the test client in phpunit test classes.  Based on those issues, and the current documentation on symfony.com, I have come up with the following code, which works for my purposes:

```
class MyWebTestCase extends WebTestCase
{
    public function setUp()
    {
        parent::setUp();
        $this->client = static::createClient();
        $this->client->disableReboot();
        $this->container = $this->client->getContainer();
    }

    public function testSomething()
    {
        ....
        $this->container->get(MyCustomService::class);  // WARNING: MyCustomService service is private, getting it from the container is deprecated since Symfony 3.2 and will fail in 4.0
        ....
    }
}

class MyCustomConsoleCommandTestCase extends WebTestCase
{
    public function setUp()
    {
        parent::setUp();
        $this->application = new Application(self::bootKernel());
        $this->client = static::createClient();
        $this->container = $this->client->getContainer();
    }

    public function testSomething()
    {
        ....
        $this->container->get(MyCustomService::class);  // WARNING: MyCustomService service is private, getting it from the container is deprecated since Symfony 3.2 and will fail in 4.0
        ....
    }
}
```

While the code works, the calls to the test client container's `get()` method produce the familiar `service is private, getting it from the container is deprecated since Symfony 3.2 and will fail in 4.0` deprecation warning.  If my approach to accessing the test container is incorrect, please let me know.  Otherwise, shouldn't `PhpUnitBridge` _not_ throw said deprecation warning in these testing contexts?

**Comment by arderyp** [comment]
Thanks for the quick response @nicolas-grekas, I see what you are saying.  I will poke around PhpUnitBridge to see if I can figure out how this would work.  If I can make progress I'll submit a PR.  Regardless, we can just leave the issue open as a reminder, as it should be implemented eventually, right?  If so, is there some 4.1 tag/milestone that should be applied?

**Comment by nicolas-grekas** [comment]
Right (but we use milestones are for PRs only). I think you can update the version in the issue description.

**Comment by nicolas-grekas** [comment]
That wouldn't work: TestContainer only exists in 4.1 but the deprecation you target is triggered by code in 3.4.
In order to silence the deprecation, we need to look at which code triggered it (the test Client).
The DI component shouldn't care at all about this totally independent concern.
Which is why I think the bridge is really the best place to do this: we don't want the deprecation to *not* be triggered. Instead, we want the bridge to intercept and silence it *starting with it's v4.1*, which is the one that can legitimately know that this case needs to be silenced.

**Comment by arderyp** [comment]
okay, thats good feedback.  I'll have to poke around PhpUnitBridge to see if I can make sense of it, and apply what you've posted here.

**Comment by arderyp** [comment]
also, @nicolas-grekas, is it acceptable to run a version of PhpUnitBridge above the version of Symfony that it is inspecting?  For example, if the Symfony code if v3.4, could one expect to be able to run PhpUnitBridge v4.1+ against it without problems?  

If not, then maybe this is all a moot point.  I say that because, if I am bound to PhpUbitBridge v3.4.x when analyzing my v3.4.x Symfony code, then I will forever see these deprications.  I would simply have to know, based on following these GitHub issues, that it's safe for me to ignore them, so long as I skip upgrading to Symfony v4.0 and wait until Symfony v4.1+.  Furthermore, if the PhpUnitBridge version must match the Symfony version, then there wouln't be any deprecation warning to intercept and suppress in PhpUnitBridge v4.1, since it would require Symfony v4.1, which has TestContainer and doesn't care about accessing private methods on the test client.

Does that make sense?  Apologies if I am misunderstanding this.

**Comment by nicolas-grekas** [comment]
That's a very good question. And the answer is yes. We actually do use the latest version of the bridge to test all versions of Symfony, from 2.7 to master. This is very unique to the bridge, and it should be stated in the documentation I suppose! (Would you mind opening a doc PR/issue?) That should be a best practice, using the latest version of the bridge no matter the major-version boundary!

**Comment by arderyp** [comment]
ahh, thats great to know.  I will open a doc PR/issue when I get a free moment tomorrow, and will continue looking into this issue here as well.  Will report back when I have more.

**Comment by arderyp** [comment]
@nicolas-grekas , regarding the documentation, see: https://github.com/symfony/symfony-docs/compare/4.0...arderyp:patch-1

**Comment by arderyp** [comment]
When installing or migrating to FLEX, the default `services_test.yml` file that gets created says:

```
services:
    _defaults:
        public: true

    # If you need to access services in a test, create an alias
    # and then fetch that alias from the container. As a convention,
    # aliases are prefixed with test. For example:
    #
    # test.App\Service\MyService: '@App\Service\MyService'
```

@nicolas-grekas (or anyone else), does the new `TestContainer` approach make the recommendation above invalid? If so, should I also submit a PR to change/remove `services_test.yml` here: https://github.com/symfony/recipes/blob/master/symfony/phpunit-bridge/3.3/config/services_test.yaml

**Comment by nicolas-grekas** [comment]
Good question. I will think we could remove it in a 4.1 variant of the recipe.

**Comment by arderyp** [comment]
@nicolas-grekas, when you have a chance, no rush.  Hoping to get feedback on the PR above from someone more familiar with the PhpUnitBridge project architecture/logic to make sure my approach above is sound.  Open to any feedback.

**Comment by stof** [comment]
I'm wondering how you can get this deprecation warning triggered by Symfony 3.4 if you are actually using symfony 4.1.

**Comment by nicolas-grekas** [comment]
Using Symfony 3.4 with the bridge 4.1, that's supported.

**Comment by arderyp** [comment]
@stof, I can say positively that it simply is not an impossible case, because my production code is triggering it.  I am running 3.4.  I am using autowiring and all of my services are private (default) and auto configured.  I followed the documentation for accessing the client/container from within tests based on the current official documentation: https://symfony.com/doc/current/testing.html#accessing-the-container

Do that, and you will trigger the warnings.

https://symfony.com/blog/new-in-symfony-4-1-simpler-service-testing

**Comment by stof** [comment]
@arderyp but this blog post is abotu a new feature in 4.1. So you are **not** using it if you are on 3.4. On 3.4, you are really accessing a private service from the main container, which is deprecated (and does not work in 4.x)

**Comment by nicolas-grekas** [comment]
@stof here is how to get this situation:
```sh
composer create-project symfony/skeleton:3.4 demo
cd demo
composer req test
```

This will install `symfony/phpunit-bridge` `4.*-stable` (see https://github.com/symfony/test-pack/pull/3)

That's why I think this is correct and should be supported.

**Comment by xabbuh** [comment]
But that does not make you use the new feature that allows you to fetch private services from the container. Thus, if your code triggers deprecations there, you should see and fix them as your tests would otherwise fail as soon as you upgrade to Symfony 4.0.

**Comment by nicolas-grekas** [comment]
Needs a rebase to account for the revert in ab09fcc2497dc254f9df256d039b656bc7b3e507

**Comment by arderyp** [comment]
@nicolas-grekas, sorry I am late catching up to this.  I just reviewed #27342.  It looks like you already made a decision on this and rolled back the bug fix we implemented in #27342.  So, should this be closed then?

**Comment by nicolas-grekas** [comment]
Issue will be solved by #27501, which will revert to return the real container here.
